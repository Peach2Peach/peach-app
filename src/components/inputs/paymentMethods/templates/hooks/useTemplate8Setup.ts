import { useCallback, useEffect, useMemo, useState } from 'react'
import { FormProps } from '../../paymentForms/PaymentMethodForm'
import { useValidatedState } from '../../../../../hooks'
import { getPaymentDataByLabel } from '../../../../../utils/account'
import { getErrorsInField } from '../../../../../utils/validation'
import { toggleCurrency } from '../../paymentForms/utils'

const referenceRules = { required: false }
const phoneRules = { required: true, phone: true, isPhoneAllowed: true }
// eslint-disable-next-line max-lines-per-function
export const useTemplate8Setup = ({
  data,
  currencies = [],
  onSubmit,
  setStepValid,
  paymentMethod,
  setFormData,
}: FormProps) => {
  const [label, setLabel] = useState(data?.label || '')
  const [phone, setPhone, phoneIsValid, phoneErrors] = useValidatedState(data?.phone || '', phoneRules)
  const [beneficiary, setBeneficiary] = useState(data?.beneficiary || '')
  const [reference, setReference, , referenceError] = useValidatedState(data?.reference || '', referenceRules)
  const [displayErrors, setDisplayErrors] = useState(false)
  const [selectedCurrencies, setSelectedCurrencies] = useState(data?.currencies || currencies)

  const onCurrencyToggle = (currency: Currency) => {
    setSelectedCurrencies(toggleCurrency(currency))
  }

  const labelRules = useMemo(
    () => ({
      required: true,
      duplicate: getPaymentDataByLabel(label) && getPaymentDataByLabel(label)!.id !== data.id,
    }),
    [data.id, label],
  )

  const labelErrors = useMemo(() => getErrorsInField(label, labelRules), [label, labelRules])

  const buildPaymentData = useCallback(
    () => ({
      id: data?.id || `${paymentMethod}-${Date.now()}`,
      label,
      type: paymentMethod,
      phone,
      beneficiary,
      reference,
      currencies: selectedCurrencies,
    }),
    [beneficiary, data?.id, label, paymentMethod, phone, reference, selectedCurrencies],
  )

  const isFormValid = useCallback(() => {
    setDisplayErrors(true)
    return phoneIsValid && labelErrors.length === 0
  }, [labelErrors.length, phoneIsValid])

  const save = () => {
    if (!isFormValid()) return

    onSubmit(buildPaymentData())
  }

  useEffect(() => {
    setStepValid(isFormValid())
    setFormData(buildPaymentData())
  }, [buildPaymentData, isFormValid, setFormData, setStepValid])

  return {
    labelInputProps: {
      onChange: setLabel,
      value: label,
      errorMessage: displayErrors ? labelErrors : undefined,
    },
    phoneInputProps: {
      onChange: setPhone,
      value: phone,
      errorMessage: displayErrors ? phoneErrors : undefined,
    },
    beneficiaryInputProps: {
      onChange: setBeneficiary,
      value: beneficiary,
      required: false,
    },
    referenceInputProps: {
      onChange: setReference,
      onSubmit: save,
      value: reference,
      errorMessage: displayErrors ? referenceError : undefined,
    },
    currencySelectionProps: {
      paymentMethod,
      selectedCurrencies,
      onToggle: onCurrencyToggle,
    },
  }
}
