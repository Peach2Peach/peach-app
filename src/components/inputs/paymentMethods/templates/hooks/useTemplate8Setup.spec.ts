import { useTemplate8Setup } from './useTemplate8Setup'
import { renderHook } from '@testing-library/react-native'

describe('useTemplate8Setup', () => {
  it('should return the correct values', () => {
    const { result } = renderHook(() =>
      useTemplate8Setup({
        data: {
          id: 'id',
          label: 'label',
          type: 'paysera',
          phone: 'phone',
          beneficiary: 'beneficiary',
          reference: 'reference',
          currencies: [],
        },
        currencies: [],
        onSubmit: jest.fn(),
        setStepValid: jest.fn(),
        paymentMethod: 'paysera',
        setFormData: jest.fn(),
      }),
    )

    expect(result.current).toMatchSnapshot()
  })
})
