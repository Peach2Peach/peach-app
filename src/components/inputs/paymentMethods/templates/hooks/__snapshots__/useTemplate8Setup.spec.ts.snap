// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTemplate8Setup should return the correct values 1`] = `
{
  "beneficiaryInputProps": {
    "onChange": [Function],
    "required": false,
    "value": "beneficiary",
  },
  "currencySelectionProps": {
    "onToggle": [Function],
    "paymentMethod": "paysera",
    "selectedCurrencies": [],
  },
  "labelInputProps": {
    "errorMessage": [],
    "onChange": [Function],
    "value": "label",
  },
  "phoneInputProps": {
    "errorMessage": [
      "please enter a valid value",
      "numbers from this country are not allowed",
    ],
    "onChange": [Function],
    "value": "phone",
  },
  "referenceInputProps": {
    "errorMessage": [],
    "onChange": [Function],
    "onSubmit": [Function],
    "value": "reference",
  },
}
`;
